Update partitioner's is_fusible heuristic to respect auto_functionalized

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134491
* __->__ #134490
* #134466
* #134364

We say Node a is fusible into node b if node b is an auto_functionalized
node that may reinplace node a later on.

This PR also changes aten.empty to be recomputable w.r.t the Partitioner
(it is, like aten.zeros, cheap to recompute and fusible into other ops).

Fixes https://github.com/pytorch/pytorch/issues/134468

Test Plan:
- new test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

zou3519 committed:
Update partitioner's is_fusible heuristic to respect auto_functionalized

We say Node a is fusible into node b if node b is an auto_functionalized
node that may reinplace node a later on.

Fixes https://github.com/pytorch/pytorch/issues/134468

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Update partitioner's is_fusible heuristic to respect auto_functionalized"


We say Node a is fusible into node b if node b is an auto_functionalized
node that may reinplace node a later on.

This PR also changes aten.empty to be recomputable w.r.t the Partitioner
(it is, like aten.zeros, cheap to recompute and fusible into other ops).

Fixes https://github.com/pytorch/pytorch/issues/134468

Test Plan:
- new test

[ghstack-poisoned]

zou3519 commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1621,6 +1621,8 @@ def is_mutation_op(node: torch.fx.Node) -> bool:
     elif node.op == "call_method":
         if _mutation_op_re.search(node.target):  # type: ignore[union-attr, arg-type]
             return True
+    if node.target is torch.ops.higher_order.auto_functionalized:
+        return False
```
auto_functionalized nodes can have `out = blah` arguments. Those are not mutable (the auto_functionalized node is always functional), so we fix that here.

Chillee replied:
This is unrelated?

zou3519 replied:
Fair enough, I'll split this out into its own PR

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -807,11 +807,29 @@ def solve_min_cut(
+            arg = b.kwargs[name]
+            if a is arg:
+                return True
+            if isinstance(arg, list):
```
should this be a tree_map or something?

zou3519 replied:
We only support TensorList args as inputs to operators, so the list is fine

Chillee commented on test/inductor/test_inplacing_pass.py:
```diff
@@ -182,6 +187,40 @@ def f(b):
+            subtest(torch.empty_like, name="empty_like"),
+        ],
+    )
+    def test_partitioner_recomputes_factory(self, factory_op):
```
I personally like `test_perf` tests, since they're "property-based" (i.e. they measure the amount of bytes read and written). It's possible we don't support HOPs yet or something though.

zou3519 replied:
I'll add some test_perf tests too

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -1273,6 +1291,8 @@ def get_default_op_list() -> OpTypes:
         aten.full,
         aten.as_strided,
         aten.zeros,
+        aten.empty,
```
I think plausibly `aten.empty` should be treated even more specially than this. Since `aten.empty` is always free to recompute, regardless of whether or not it's fusible into a downstream op.

zou3519 committed:
Update on "Update partitioner's is_fusible heuristic to respect auto_functionalized"


We say Node a is fusible into node b if node b is an auto_functionalized
node that may reinplace node a later on.

This PR also changes aten.empty to be recomputable w.r.t the Partitioner
(it is, like aten.zeros, cheap to recompute and fusible into other ops).

Fixes https://github.com/pytorch/pytorch/issues/134468

Test Plan:
- new test

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
